File: app/src/main/java/com/castelijns/mmdemo/photos/PhotosFragment.java
package com.castelijns.mmdemo.photos;

import android.app.Activity;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v7.widget.GridLayoutManager;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.util.DisplayMetrics;
import android.util.SparseArray;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.castelijns.mmdemo.R;
import com.castelijns.mmdemo.app.BaseFragment;
import com.castelijns.mmdemo.models.Photo;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;

import butterknife.BindView;
import io.github.luizgrp.sectionedrecyclerviewadapter.SectionedRecyclerViewAdapter;

public class PhotosFragment extends BaseFragment implements PhotosContract.View {

    public static final int PHOTOS_GRID_COLS = 5;

    @BindView(R.id.rv_photos)
    RecyclerView rvPhotos;

    private PhotosContract.Presenter presenter;

    private SectionedRecyclerViewAdapter sectionAdapter;
    private int itemWidth;
    private PhotosAdapter adapter;
    private List<Photo> photos;

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        // Calculate how wide one col is allowed to be.
        DisplayMetrics displaymetrics = new DisplayMetrics();
        getActivity().getWindowManager().getDefaultDisplay().getMetrics(displaymetrics);
        itemWidth = displaymetrics.widthPixels / PHOTOS_GRID_COLS;

        presenter = new PhotosPresenter(this);
    }

    @Nullable
    @Override
    public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        return inflater.inflate(R.layout.fragment_photos, container, false);
    }

    @Override
    public void onViewCreated(View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);

        sectionAdapter = new SectionedRecyclerViewAdapter();

        GridLayoutManager gridLayoutManager = new GridLayoutManager(getContext(), PHOTOS_GRID_COLS);
        gridLayoutManager.setSpanSizeLookup(new GridLayoutManager.SpanSizeLookup() {
            @Override
            public int getSpanSize(int position) {
                switch(sectionAdapter.getSectionItemViewType(position)) {
                    case SectionedRecyclerViewAdapter.VIEW_TYPE_HEADER:
                        return PHOTOS_GRID_COLS;
                    default:
                        return 1;
                }
            }
        });

        rvPhotos.setLayoutManager(gridLayoutManager);

        presenter.start();
    }

    @Override
    public void showPhotos(SparseArray<List<Photo>> albumPhotos) {
        for (int i = 0; i < albumPhotos.size(); i++) {
            sectionAdapter.addSection(new PhotoSection(getActivity(), albumPhotos.keyAt(i),
                    albumPhotos.valueAt(i), itemWidth));
        }

        rvPhotos.setAdapter(sectionAdapter);
    }
}

File: app/src/main/kotlin/com/castelijns/mmdemo/photos/PhotosFragment.kt
package com.castelijns.mmdemo.photos

import android.content.Intent
import android.os.Bundle
import android.util.DisplayMetrics
import android.util.SparseArray
import android.view.LayoutInflater
import android.view.Menu
import android.view.View
import android.view.ViewGroup
import android.widget.EditText
import android.widget.ImageView
import androidx.appcompat.app.AlertDialog
import androidx.core.app.ActivityOptionsCompat
import androidx.core.view.ViewCompat
import androidx.recyclerview.widget.GridLayoutManager
import com.castelijns.mmdemo.MainActivity
import com.castelijns.mmdemo.R
import com.castelijns.mmdemo.app.BaseListFragment
import com.castelijns.mmdemo.models.Album
import com.castelijns.mmdemo.models.Photo
import com.castelijns.mmdemo.photodetail.PhotoDetailActivity
import io.github.luizgrp.sectionedrecyclerviewadapter.SectionedRecyclerViewAdapter
import kotlinx.android.synthetic.main.fragment_photos.*
import java.util.*

class PhotosFragment : BaseListFragment(), PhotosContract.View {

    private var presenter: PhotosPresenter? = null

    private var sectionAdapter: SectionedRecyclerViewAdapter? = null
    private var itemWidth: Int = 0

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setHasOptionsMenu(true)

        // Calculate how wide one col is allowed to be.
        val displaymetrics = DisplayMetrics()
        activity!!.windowManager.defaultDisplay.getMetrics(displaymetrics)
        itemWidth = displaymetrics.widthPixels / PHOTOS_GRID_COLS

        presenter = PhotosPresenter(this)
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        return inflater.inflate(R.layout.fragment_photos, container, false)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        sectionAdapter = SectionedRecyclerViewAdapter()

        val gridLayoutManager = GridLayoutManager(getContext(), PHOTOS_GRID_COLS)
        gridLayoutManager.spanSizeLookup = object : GridLayoutManager.SpanSizeLookup() {
            override fun getSpanSize(position: Int): Int {
                return when (sectionAdapter!!.getSectionItemViewType(position)) {
                    SectionedRecyclerViewAdapter.VIEW_TYPE_HEADER -> PHOTOS_GRID_COLS
                    else -> 1
                }
            }
        }

        rv_photos.layoutManager = gridLayoutManager

        // If args were passed, presenter should start by doing something different from default.
        val args = arguments
        if (args != null) {
            val albumId = arguments!!.getInt(MainActivity.EXTRA_ALBUM_ID)
            presenter!!.start(albumId)
        } else {
            presenter!!.start()
        }
    }

    override fun onDestroyView() {
        presenter!!.stop()
        super.onDestroyView()
    }

    override fun onPrepareOptionsMenu(menu: Menu?) {
        menu!!.findItem(R.id.action_filter).isVisible = true
        super.onPrepareOptionsMenu(menu)
    }

    fun filterActionClicked() {
        val view = layoutInflater.inflate(R.layout.dialog_input, null)
        AlertDialog.Builder(getContext()!!)
                .setTitle(R.string.filter_albums_title)
                .setMessage(R.string.filter_albums_message)
                .setView(view)
                .setPositiveButton(R.string.filter) { dialog, which ->
                    val albumId = Integer.parseInt((view.findViewById<View>(R.id.et_input) as EditText)
                            .text.toString())
                    presenter!!.filterAlbums(albumId)
                }
                .setNegativeButton(R.string.cancel, null)
                .show()
    }

    override fun clearList() {
        sectionAdapter!!.removeAllSections()
    }

    override fun showPhotosByAlbum(photosByAlbum: Map<Album, List<Photo>>) {
        photosByAlbum.keys.forEach { album ->
            val section = PhotoSection(activity!!, album.id, album.title, photosByAlbum[album]!!,
                    itemWidth)

            section.setItemClickListener(object : PhotoSection.ItemClickListener {
                override fun onClick(photo: Photo, ivPhoto: ImageView) {
                    presenter!!.onPhotoClicked(photo, ivPhoto)
                }
            })

            sectionAdapter?.addSection(section)
        }

        rv_photos.adapter = sectionAdapter
    }

    override fun showPhotoCount(photoCount: Int, albumCount: Int) {
        setHeaderText(String.format(Locale.getDefault(),
                "%d %s in %d %s", photoCount, getString(R.string.header_photos), albumCount,
                getString(R.string.header_albums)))
    }

    override fun showPhotoDetail(photo: Photo, ivPhoto: ImageView) {
        val transitionName = ViewCompat.getTransitionName(ivPhoto)

        val intent = Intent(getContext(), PhotoDetailActivity::class.java)
        intent.putExtra(EXTRA_PHOTO, photo)
        intent.putExtra(EXTRA_PHOTO_TRANSITION, transitionName)

        val options = ActivityOptionsCompat.makeSceneTransitionAnimation(
                activity!!, ivPhoto, transitionName!!)

        startActivity(intent, options.toBundle())
    }

    companion object {

        const val PHOTOS_GRID_COLS = 5

        const val EXTRA_PHOTO = "extra_photo"
        const val EXTRA_PHOTO_TRANSITION = "extra_photo_transition"
    }
}
